capirca/lib/versa.py

```diff
@@ -0,0 +1,877 @@
+        child.PrintTreeInt(num+1)
+    if len(self.name) > 0:
+      self.target.append(f'{self.INDENT*num}' + '}')
+
+  def ResetTarget(self):
+    """Reset the target """
+    self.target.clear()
+
+class Term(aclgenerator.Term):
+  """Representation of an individual Versa term.
+
+     This is mostly useful for the __str__() method.
+
+     Args:
+       obj: a policy.Term object
+       filter_options: list of remaining target options (zones)
+  """
+
+  ACTIONS = {'accept': 'permit',
+             'deny': 'deny',
+             'reject': 'reject',
+             'count': 'count',
+             'log': 'log',
+             'dscp': 'dscp'}
+
+  def __init__(self, term, from_zone, to_zone, addrbook, verbose=True):
+    super().__init__(term)
+    self.term = term
+    self.from_zone = []
+    self.to_zone = []
+    self.verbose = verbose
+    self.addrbook = addrbook
+    self.app = []
+
+    if term.source_zone:
+      self.from_zone = term.source_zone
+    elif from_zone != 'all':
+      self.from_zone.append(from_zone)
+
+    if term.destination_zone:
+      self.to_zone = term.destination_zone
+    elif to_zone != 'all':
+      self.to_zone.append(to_zone)
+
+  def AddApplication(self,new_app):
+    """Add new app"""
+    self.app.append(new_app)
+
+
+  def BuildTermZone(self, p_node, zonetype):
+
+    if zonetype == 'src':
+      mtype = self.from_zone
+      match_str = 'source'
+      maddr = self.term.source_address
+      maddr_ex = self.term.source_address_exclude
+      addr_str = 'src-'
+    else:
+      mtype = self.to_zone
+      match_str = 'destination'
+      maddr = self.term.destination_address
+      maddr_ex = self.term.destination_address_exclude
+      addr_str = 'dest-'
+
+
+    zone_str = '['
+    for zone in mtype:
+      zone_str = zone_str + ' ' + zone
+    zone_str = zone_str + ' ];'
+    match_node= Tree(match_str)
+    match_node.AddParent(p_node)
+    matchnode_zone = Tree('zone', 'zone-list ' + zone_str)
+    matchnode_zone.AddParent(match_node)
+
+    if maddr:
+      if addr_str + self.term.name in self.addrbook:
+        addr_list = ['address-group-list [ ' + addr_str + self.term.name+ ' ];']
+        if maddr_ex:
+          addr_list.append = 'negate;'
+        addr_t = Tree('address',addr_list)
+        addr_t.AddParent(matchnode_zone )
+      else:
+        pass
+
+  def BuildTermApp(self, p_node):
+    """Build Term app"""
+    mstr = []
+    if self.app:
+      apps = list(filter(lambda x: x['name'] == self.term.name, self.app))
+      if len(apps) > 0:
+        slist = ''
+        for i in range(0,len(apps)):
+          for j in range(0,len(apps[i]['protocol'])):
+            slist = slist + ' ' + self.term.name + '-app' + str(j+1)
+        if len(slist) > 0:
+          slist = 'services-list [' + slist + ' ];'
+          mstr.append(slist)
+
+    if self.term.versa_application:
+      predef_str = 'predefined-services-list ['
+      for predef in self.term.versa_application:
```

Comment:
We probably need to do some validation checking here on what valid versa applications are.

Reply:
versa.application is a list and it is check in Line 237